File: common/src/main/java/allyouneed/mixin/ACraftingTreeProcess.java
package allyouneed.mixin;

import appeng.api.config.Actionable;
import appeng.api.crafting.IPatternDetails;
import appeng.api.networking.crafting.ICraftingService;
import appeng.api.stacks.AEKey;
import appeng.api.stacks.GenericStack;
import appeng.api.stacks.KeyCounter;
import appeng.crafting.CraftBranchFailure;
import appeng.crafting.inv.CraftingSimulationState;
import allyouneed.pattern.adaptive.AdaptiveStatisticalPattern;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Custom CraftingTreeProcess that intercepts adaptive statistical patterns
 * and scales their inputs by ceil(N/p) instead of using binomial distribution.
 */
public class ACraftingTreeProcess {

    private final ACraftingTreeNode parent;
    final IPatternDetails details;
    private final ACraftingCalculation job;
    private final Map<ACraftingTreeNode, Long> nodes = new LinkedHashMap<>();
    boolean possible = true;
    private boolean containerItems;
    private boolean limitQty;

    public ACraftingTreeProcess(ICraftingService cc, ACraftingCalculation job,
                                IPatternDetails details,
                                ACraftingTreeNode craftingTreeNode) {
        this.parent = craftingTreeNode;
        this.details = details;
        this.job = job;

        updateLimitQty();

        final IPatternDetails.IInput[] inputs = this.details.getInputs();
        for (int x = 0; x < inputs.length; ++x) {
            var input = inputs[x];
            var firstInput = input.getPossibleInputs()[0];
            this.nodes.put(new ACraftingTreeNode(cc, job, firstInput.what(), firstInput.amount(), this, x),
                    input.getMultiplier());
        }
    }

    boolean notRecursive(IPatternDetails details) {
        return this.parent == null || this.parent.notRecursive(details);
    }

    private void updateLimitQty() {
        for (IPatternDetails.IInput input : details.getInputs()) {
            var primaryInput = input.getPossibleInputs()[0];
            boolean isAnInput = false;

            for (var output : details.getOutputs()) {
                if (output.what().matches(primaryInput)) {
                    isAnInput = true;
                    break;
                }
            }

            if (isAnInput) {
                this.limitQty = true;
            }

            if (input.getRemainingKey(primaryInput.what()) != null) {
                this.limitQty = this.containerItems = true;
            }
        }
    }

    boolean limitsQuantity() {
        return this.limitQty;
    }

    void request(CraftingSimulationState inv, long times)
            throws CraftBranchFailure, InterruptedException {
        this.job.handlePausing();

        var containerItems = this.containerItems ? new KeyCounter() : null;

        for (var entry : this.nodes.entrySet()) {
            entry.getKey().request(inv, entry.getValue() * times, containerItems);
        }

        if (containerItems != null) {
            for (var stack : containerItems) {
                inv.insert(stack.getKey(), stack.getLongValue(), Actionable.MODULATE);
                inv.addStackBytes(stack.getKey(), stack.getLongValue(), 1);
            }
        }

        for (var out : this.details.getOutputs()) {
            inv.insert(out.what(), out.amount() * times, Actionable.MODULATE);
        }

        inv.addCrafting(details, times);
        inv.addBytes(times);
    }

    long getNodeCount() {
        long tot = 0;
        for (ACraftingTreeNode node : this.nodes.keySet()) {
            tot += node.getNodeCount();
        }
        return tot;
    }

    long getOutputCount(AEKey what) {
        long tot = 0;
        for (var is : this.details.getOutputs()) {
            if (what.matches(is)) {
                tot += is.amount();
            }
        }
        return tot;
    }

    boolean hasMultiplePaths() {
        for (var entry : nodes.entrySet()) {
            if (entry.getKey().hasMultiplePaths()) {
                return true;
            }
        }
        return false;
    }

    double getSuccessProbability() {
        double ownProb = 1.0;
        if (this.details instanceof AdaptiveStatisticalPattern asp) {
            ownProb = 1.0 - (1.0 - asp.getProbability());
        }
        double childProb = 1.0;
        for (var entry : this.nodes.entrySet()) {
            childProb *= entry.getKey().getSuccessProbability();
        }
        return ownProb * childProb;
    }
}

File: common/src/main/kotlin/allyouneed/logic/crafting/ACraftingTreeProcess.kt
package allyouneed.logic.crafting

import allyouneed.pattern.adaptive.AdaptiveStatisticalPattern
import appeng.api.config.Actionable
import appeng.api.crafting.IPatternDetails
import appeng.api.networking.crafting.ICraftingService
import appeng.api.stacks.AEKey
import appeng.api.stacks.KeyCounter
import appeng.crafting.CraftBranchFailure
import appeng.crafting.inv.CraftingSimulationState

/**
 * Custom CraftingTreeProcess that intercepts adaptive statistical patterns
 * and scales their inputs by ceil(N/p) instead of using binomial distribution.
 */
class ACraftingTreeProcess(
    cc: ICraftingService,
    private val job: ACraftingCalculation,
    val details: IPatternDetails,
    private val parent: ACraftingTreeNode?
) {
    private val nodes: MutableMap<ACraftingTreeNode, Long?> = LinkedHashMap<ACraftingTreeNode, Long?>()
    var possible: Boolean = true
    private var containerItems = false
    private var limitQty = false

    init {
        updateLimitQty()

        val inputs = this.details.inputs
        for (x in inputs.indices) {
            val input = inputs[x]
            val firstInput = input.possibleInputs[0]
            this.nodes[ACraftingTreeNode(cc, job, firstInput.what(), firstInput.amount(), this, x)] = input.multiplier
        }
    }

    fun notRecursive(details: IPatternDetails): Boolean {
        return this.parent == null || this.parent.notRecursive(details)
    }

    private fun updateLimitQty() {
        for (input in details.inputs) {
            val primaryInput = input.possibleInputs[0]
            var isAnInput = false

            for (output in details.outputs) {
                if (output.what().matches(primaryInput)) {
                    isAnInput = true
                    break
                }
            }

            if (isAnInput) {
                this.limitQty = true
            }

            if (input.getRemainingKey(primaryInput.what()) != null) {
                this.containerItems = true
                this.limitQty = true
            }
        }
    }

    fun limitsQuantity(): Boolean {
        return this.limitQty
    }

    @Throws(CraftBranchFailure::class, InterruptedException::class)
    fun request(inv: CraftingSimulationState, times: Long) {
        this.job.handlePausing()

        val containerItems = if (this.containerItems) KeyCounter() else null

        for (entry in this.nodes.entries) {
            entry.key.request(inv, entry.value!! * times, containerItems)
        }

        if (containerItems != null) {
            for (stack in containerItems) {
                inv.insert(stack.key, stack.longValue, Actionable.MODULATE)
                inv.addStackBytes(stack.key, stack.longValue, 1)
            }
        }

        for (out in this.details.outputs) {
            inv.insert(out.what(), out.amount() * times, Actionable.MODULATE)
        }

        inv.addCrafting(details, times)
        inv.addBytes(times.toDouble())
    }

    val nodeCount: Long
        get() {
            var tot: Long = 0
            for (node in this.nodes.keys) {
                tot += node.nodeCount
            }
            return tot
        }

    fun getOutputCount(what: AEKey): Long {
        var tot: Long = 0
        for (`is` in this.details.outputs) {
            if (what.matches(`is`)) {
                tot += `is`.amount()
            }
        }
        return tot
    }

    fun hasMultiplePaths(): Boolean {
        for (entry in nodes.entries) {
            if (entry.key.hasMultiplePaths()) {
                return true
            }
        }
        return false
    }

    val successProbability: Double
        get() {
            var ownProb = 1.0
            if (this.details is AdaptiveStatisticalPattern) {
                ownProb = 1.0 - (1.0 - details.probability)
            }
            var childProb = 1.0
            for (entry in this.nodes.entries) {
                childProb *= entry.key.successProbability
            }
            return ownProb * childProb
        }
}
